crates/matrix-sdk-ui/src/timeline/inner.rs

```diff
@@ -142,6 +168,119 @@ impl<P: RoomDataProvider> TimelineInner<P> {
+                let txn_id = match in_flight {
+                    Some(ReactionState::Sending(txn_id)) => {
+                        // Use the transaction ID as the in flight request
+                        txn_id.clone()
+                    }
+                    _ => txn_id,
+                };
+
+                let event_content = AnyMessageLikeEventContent::Reaction(
+                    ReactionEventContent::from(annotation.clone()),
+                );
+                self.handle_local_event_internal(
+                    &mut state,
+                    sender,
+                    sender_profile,
+                    txn_id.clone(),
+                    event_content.clone(),
+                )
+                .await;
+                ReactionState::Sending(txn_id)
+            }
+            (to_redact_local, to_redact_remote) => {
+                // The reaction exists, redact local echo and/or remote echo
+                let no_reason = RoomRedactionEventContent::default();
+                self.handle_local_redaction_internal(
+                    &mut state,
+                    sender,
+                    sender_profile,
+                    txn_id.clone(),
+                    (to_redact_local.cloned(), to_redact_remote.cloned()),
+                    no_reason.clone(),
+                )
+                .await;
+
+                // Remember the remote echo to redact on the homeserver
+                ReactionState::Redacting(to_redact_remote.cloned())
+            }
+        };
+
+        state.reaction_state.insert(annotation.into(), reaction_state.clone());
+
+        // Check the action to perform depending on any in flight request
+        let in_flight = state.in_flight_reaction.get::<AnnotationKey>(&annotation.into());
+        let result = match in_flight {
+            Some(_) => {
+                // There is an in-flight request
+                // This local reaction should not initiate a new request
+                ReactionAction::None
+            }
+
+            None => {
+                // There is no in-flight request
+                match reaction_state.clone() {
+                    ReactionState::Redacting(event_id) => match event_id {
+                        Some(event_id) => ReactionAction::RedactRemote(event_id),
+                        // If we just redacted a local echo, no need to send a remote request
+                        None => ReactionAction::None,
+                    },
+                    ReactionState::Sending(txn_id) => ReactionAction::SendRemote(txn_id),
+                }
+            }
+        };
+
+        match result {
+            ReactionAction::None => {}
+            ReactionAction::SendRemote(_) | ReactionAction::RedactRemote(_) => {
+                // Remember the new in flight request
+                state.in_flight_reaction.insert(annotation.into(), reaction_state);
+            }
+        };
```

Comment:
nit: This `match` could be avoided if we duplicated line 277 in the two places in the above match. I would prefer this slightly better, but it's personal taste, so feel free to dismiss :)

Reply:
I added two matches to make it a bit clearer what is being done (i.e. first calculate the result, then update the state) and avoid the side effect getting lost amongst the other logic.

Given the function is already pretty long I'd probably err on the side of clarity if that makes sense to you also?